### PR TITLE
Make the set language more future proof

### PR DESCRIPTION
**Pull request by diml**
This PR will allow us to add support for things like `(if ...)` in ordered sets in the future without breaking compatibility:

- it forbids list starting by an atom not starting with `-` or `:` in dune files
- since we have many of these for `modules`, I also made a simple change to allow to remove the toplevel parentheses, which is nice on its own

### CODE REVIEW

File: src/ordered_set_lang.ml [hunk 1]
```diff
@@ -45,19 +45,30 @@ module Parse = struct
...
-      | List (_, Atom (_, A ":include") :: _) -> inc
-      | List _ -> enter (many [])
-    and many acc =
+      | List (_, Atom (loc, A s) :: _) -> begin
+          match s, kind with
+          | ":include", _ -> inc
+          | s, Dune when s <> "" && s.[0] <> '-' && s.[0] <> ':' ->
+            Loc.fail loc
+              "This atom must be quoted because it is the first element \
+               of a list and doesn't start with - or :"
```
Comment (emillon) [inline src/ordered_set_lang.ml]: Can you document this syntax in the "Ordered set language" section in doc/jbuild.rst?

File: bin/dune [hunk 2]
```diff
@@ -1,20 +1,20 @@
 (library
  (name      main)
- (modules   (main))
```
Comment (emillon) [inline bin/dune]: This also applies to `(flags)` I guess - should we update the tests to remove the parentheses in these as well?

**Comment by rgrinberg** [comment]
Does this still allow optional parens in the OSL? I think that those might still be useful for for being precise with the set operations. E.g. being  able to do `((x \ a b c) a)`

File: src/ordered_set_lang.ml (hunk 1 above)
Comment (diml) [reply src/ordered_set_lang.ml]: done

File: bin/dune (hunk 2 above)
Comment (diml) [reply bin/dune]: Yes, I only fixed the build failures so far, but we could indeed update all the `flags` fields.

**Comment by diml** [comment]
@rgrinberg yes, but you have to quote the first element of lists when they don't start with `-` or `:`: `(("x" \ a b c) a)`. In pratice that shouldn't be an issue though, since `\` is pretty much only used after `:standard` which must not be quoted.

File: bin/dune (hunk 2 above)
Comment (emillon) [reply bin/dune]: Ah ok I didn't realise this was mandatory. IIUIC, an alternative would have been `(modules ("main"))`?

Comment (diml) [reply bin/dune]: yep

**Pull request merged**